networkx/algorithms/bipartite/matching.py

```diff
@@ -582,3 +584,78 @@ def minimum_weight_full_matching(G, top_nodes=None, weight="weight"):
     # add the ones from right to left as well.
     d.update({v: u for u, v in d.items()})
     return d
+
+
+def dulmage_mendelsohn_decomposition(G, top_nodes, matching=None):
+    """Return a partition of the nodes of ``G`` into the subsets defined by
+    the Dulmage-Mendelsohn decomposition
+
+    The Dulmage-Mendelsohn decomposition [1]_ partitions nodes of a bipartite
+    graph into unique subsets that determine all nodes that can possibly be
+    unmatched in some maximum matching. The subsets are defined in terms of
+    the bipartite sets of the graph. They are:
+
+    - The unmatched nodes in one set, and all nodes reachable via an alternating
+      path from them
+    - The unmatched nodes in the other set, and all nodes reachable via an
+      alternating path from them
+    - The nodes that are unreachable via alternating path from any unmatched
+      node
+
+    Unmatched nodes and alternating paths are defined with respect to a
+    particular maximum matching. However, the subsets are independent of the
+    matching used to compute them.
+
+    In contrast to the other functions in this module, ``top_nodes`` is a
+    required argument, even if there is only one possible bipartition. This
+    is so there is never ambiguity about which set of nodes each "reachable"
+    subset is reachable from.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    top_nodes : container
+
+    Returns
+    -------
+    reachable_from_top : list
+
+        List of nodes reachable by an alternating path from unmatched top nodes
+
+    reachable_from_bot : list
+
+        List of nodes reachable by an alternating path from unmatched bottom
+        nodes
+
+    unreachable : list
+
+        List of nodes that are not reachable by an alternating path from
+        unmatched nodes.
+
+    References
+    ----------
+    .. [1] Dulmage and Mendelsohn. "Coverings of Bipartite Graphs". Can. J. Math.,
+        1958.
+
+    """
+    if not is_bipartite(G):
+        raise nx.NetworkXError("Provided graph is not bipartite")
+    elif not is_bipartite_node_set(G, top_nodes):
+        raise nx.NetworkXError("Provided nodes are not a valid bipartite set")
+    if matching is None:
+        matching = maximum_matching(G, top_nodes=top_nodes)
+    top_set = set(top_nodes)
+    bot_nodes = [node for node in G if node not in top_set]
+
+    unmatched_top = [node for node in top_nodes if node not in matching]
+    unmatched_bot = [node for node in bot_nodes if node not in matching]
```

Comment:
Thanks for the PR! 
I haven't had a chance to go through the whole thing, but I saw this so am leaving a note.

Just like `union` there is a set operation called `difference` which would replace the list comprehensions that check `if node not in matching`. It's convenient to write as:
```python
unmatched_top = top_nodes - matching.keys()
```
It might even work without the `.keys()` part, but that depends on your python version.
Similar operations exist for `set_reachable_from_top & G.nodes` for intersection. 
I think `|` is the symbol for union.  There are functions for each of these operations like the `union` you are using.

Reply:
Just updated to use set operators in a few places. I could also remove
```python
    unreachable = [node for node in G if node not in set_reachable]
    reachable_from_top = [node for node in G if node in set_reachable_from_top]
    reachable_from_bot = [node for node in G if node in set_reachable_from_bot]
```
and just set `unreachable = G.nodes - set_reachable`, then
```python
return set_reachable_from_top, set_reachable_from_bot, unreachable
```
The reason I went with the list comprehensions is to ensure that the returned containers have a unique order (determined by the order in `G`, which I assume is uniquely defined - correct me if I'm wrong). My understanding is that set operations, including those involving `G.nodes` such as
```python
unreachable = G.nodes - set_reachable
reachable_from_top = G.nodes & set_reachable_from_top
```
do not have a uniquely defined order as the resulting objects are sets.

Let me know if you think this is worth designing around. My first instinct would be to preserve order and return lists, but I see that some NX functions, e.g. `bipartite.sets`, return sets rather than ordered containers.

Reply:
The order is not useful (to my knowledge) for the partition itself, but I've found it can be useful for debugging. However, any calling routine can easily use these sets to construct lists with the same elements and a well-defined order, as you suggest.